src/components/input/multimodal/uploader/uploader.tsx

```diff
@@ -0,0 +1,304 @@
+import './uploader.css'
+
+import Box from '@mui/material/Box'
+import Card from '@mui/material/Card'
+import IconButton from '@mui/material/IconButton'
+import LinearProgress from '@mui/material/LinearProgress'
+import Typography from '@mui/material/Typography'
+import type { AxiosProgressEvent } from 'axios'
+import axios from 'axios'
+import React, { useEffect, useState } from 'react'
+import { createPortal } from 'react-dom'
+import { v4 as getUUID } from 'uuid'
+
+import { shortenString } from '../../../helper'
+import type { UploaderProps } from '../../../types'
+
+const maximumFileNameLength = 15
+const maximumLoadingProgress = 100
+
+interface FileInfo {
+  id: string
+  name: string
+  loadingProgress: number
+  abortController: AbortController
+}
+
+export function getFilesToAdd(
+  files: File[],
+  totalFileCount: number,
+  maxFileCount?: number
+): File[] | undefined {
+  if (maxFileCount) {
+    if (totalFileCount <= maxFileCount) {
+      return files
+    } else {
+      return files.slice(0, maxFileCount - totalFileCount)
+    }
+  } else {
+    return files
+  }
+}
+
+export function getFileSizeAbbrev(bytes: number): string {
+  const units = ['Bytes', 'KB', 'MB', 'GB']
+  let unitIndex = 0
+
+  const byteConversionRate = 1024
+  while (bytes >= byteConversionRate && unitIndex < units.length - 1) {
+    bytes /= byteConversionRate
+    unitIndex++
+  }
+
+  const formattedString = bytes.toFixed(1).toString().replace(/\.0$/, '')
+  return `${formattedString} ${units[unitIndex]}`
+}
+
+function Uploader(props: UploaderProps) {
+  const [addedFiles, setAddedFiles] = useState<FileInfo[]>([])
+  const [errorMessages, setErrorMessages] = useState<string[]>([])
+  const inputId = getUUID()
+
+  useEffect(() => {
+    setAddedFiles([])
+    setErrorMessages([])
+  }, [props.messageId])
+
+  function rejectFile(fileName: string) {
+    setErrorMessages((prevMessages) => [
+      ...prevMessages,
+      `Failed to upload ${fileName}. You cannot upload files larger than ${
```

Comment:
In a follow up, maybe we can add use `shortenString` here as well

Reply:
Actually I think full name would be better. I shortened the string in the `filePreview` because I want to have consistent sizes for `filePreview`. But for error messages, long name doesn't hurt and if people have files with the same prefix, they won't be confused about which file is not uploaded successfully.